docs/adr/010_device_request_queue.md

```diff
@@ -0,0 +1,154 @@
+be executed in a serial/exclusive fashion on the `LoRaDevice` itself. This
+will ensure that there is only one operation acting on the connection at any
+given time. Making each execution exclusive (think queue) gives a
+deterministic way of adding a close operation without affecting any other
+operations that may be in flight.
+
+Unfortunately, the initialization of a `LoRaDevice` cannot be addressed with
+the same approach since it requires a two-step initialization per its current
+design. That is, a `LoRaDevice` is _minimally initialized_ when created so its
+`InitializeAsync` method can be called to fetch the twin information. Until
+that completes, a `LoRaDevice` is not considered technically ready for being
+put in the cache. Since a `LoRaDevice` cannot be discovered during its
+initialization, it cannot be synchronized with the other operations. To work
+around this,m load operations will be synchronized in the `LoRaDeviceRegistry`
+to ensure there is only ever one operation loading the twins for a particular
+DevEUI. It was also decided that closing the connection after initialization
+is the easiest approach to delay the connection ownership decision to after
+the arrival of the first message.
+
+The approach to simplify the entire flow and processing of a message into a
+simple request-response model will be taken a stretch goal so that the
+connection management can be implemented and tested earlier. A quick spike
+demonstrated that the changes to the main code base would be fairly contained
+and the largest impact is expected to be in adapting the tests (which could
+also be done with a stop-gap measure where the tests are adapted after the
+initial refactoring of the code base).
+
+Based on a spike of the changes the simple request-response model would
+require, it seems plausible to achieve the refactoring within (at most) two
+weekly sprints. It is expected that the bulk of the time will be spent in
+refactoring the test code.
+
+## Consequences
+
+The chosen approach has the following benefits:
+
+- Single use of `Task.Run` in `LnsProtocolMessageProcessor`. This reduces the
+  number execution forks to consider as well as chances of exceptions going
+  unobserved.
+
+- Removes many abstractions like `ILoRaDeviceRequestQueue` and
+  `IMessageDispatcher`, implementations like `DeviceLoaderSynchronizer` and
+  `ExternalGatewayLoRaRequestQueue`, and potentially more.
+
+- Easier to reason about the overall message flow.
+
+- Use of the regular async-await programming model, including error-handling
+  and cancellation that's built-in into tasks.
+
+- Transparent use of queues through tasks in `LoRaDevice` to serialize
+  execution. This subsumes the first approach (discussed in the introductory
+  section) by having implicit rather explicit queuing (at the
+  application-level).
+
+- Greatest potential to simplify tests since assertions can rely on simply
+  return values and exceptions.
```

Comment:
```suggestion
  return values and exceptions, instead of success/failure notifications.
```

The reader might not know what the current status is for assertions. What do you think about writing it down?